---
title: Analysis of MC38-HER2 CAR-T cells, with or without IL10 expression  
author: "M. Andreatta <massimo.andreatta at unil.ch> and S. Carmona <santiago.carmona at unil.ch>"
knit: (function(input_file, encoding) {
  out_dir <- 'docs';
  rmarkdown::render(input_file,
 encoding=encoding,
 output_file=file.path(dirname(input_file), out_dir, 'index.Il10_CART.html'))})
#output: html_notebook
---

```{r}
renv::restore()

library(remotes)
library(Seurat)
library(ggplot2)
#remotes::install_github("carmonalab/UCell")
#remotes::install_github("carmonalab/scGate")
#remotes::install_github("ncborcherding/escape@dev")

library(escape)
library(viridis)
```

Set up directories
```{r}
#NB: download the data and set 'ddir' to your data directory
ddir <- "~/Dropbox/CSI/Collaborations/LiTang"
  
#make plots directory
dir.create(file.path("plots"), showWarnings = FALSE)
```

```{r}
data.seurat <- Read10X(sprintf("%s/data/HER2_CAR_T", ddir))

#These are multiplexed data. Separate samples by their tag
names(data.seurat)
gex <- CreateSeuratObject(data.seurat[["Gene Expression"]], project="IL10_CART", min.cells = 3, min.features = 100)
gex 

valid.hashtags <- c("CMO309","CMO310")
valid.cells <- colnames(gex)
mpc <- data.seurat[["Multiplexing Capture"]][valid.hashtags,valid.cells]

gex[["hash"]] <- CreateAssayObject(counts = mpc)

gex <- NormalizeData(gex, assay = "hash", normalization.method = "CLR")

gex <- HTODemux(gex, assay = "hash", positive.quantile = 0.99)
```

Visualize demux results
```{r}
table(gex$hash.ID)
Idents(gex) <- "hash.ID"
RidgePlot(gex, assay = "hash", features = rownames(gex[["hash"]])[1:2], ncol = 2)
```

See demux assignments by CellRanger. Read these assignments from the provided metadata (multi folder)
```{r}
demux.CR <- read.csv(sprintf("%s/multi/multiplexing_analysis/tag_calls_per_cell.csv",ddir), header=T)
rownames(demux.CR) <- demux.CR$cell_barcode

cells.use <- intersect(colnames(gex), demux.CR$cell_barcode)

gex@meta.data[cells.use,"sample.CR"] <- demux.CR[cells.use, "feature_call"]

table(gex$sample.CR, gex$hash.ID, useNA = "ifany")
gex
```


Visualize demux results
```{r}
RidgePlot(gex, assay = "hash", features = rownames(gex[["hash"]])[1:2], ncol = 2)
```

Compare demuxing done above from scratch (using both hashtags) to demux by CellRanger
```{r}
FeatureScatter(gex, feature1 = "hash_CMO309", feature2 = "hash_CMO310", group.by = "sample.CR") + ggtitle("By cellranger demux")
FeatureScatter(gex, feature1 = "hash_CMO309", feature2 = "hash_CMO310", group.by = "hash.ID") + ggtitle("By percentiles")
```

Which annotation to use - automated from cell ranger, or based on HTODemux
```{r}
use.cellranger <- F

if (use.cellranger) {
   gex$Sample <- gex$sample.CR
} else {
  gex$Sample <- gex$hash.ID
}

Idents(gex) <- "Sample"
```


Ribosomal and mitochondrial content
```{r}
data.seurat <- gex
data.seurat <- AddMetaData(data.seurat, metadata = PercentageFeatureSet(data.seurat, pattern = "^Rp[ls]"), col.name = "percent.ribo")
data.seurat <- AddMetaData(data.seurat, metadata = PercentageFeatureSet(data.seurat, pattern = "^mt-"), col.name = "percent.mito")
```


```{r}
Idents(data.seurat) <- "Sample"
VlnPlot(data.seurat, features = c("nFeature_RNA", "nCount_RNA","percent.ribo","percent.mito"), ncol = 2, pt.size=0)
```

Filter out outliers & low quality cells
```{r}
cutoffs <- list()
cutoffs[["percent.ribo"]] <- c(min=max(quantile(data.seurat$percent.ribo,probs=c(0.01)),0),max=min(quantile(data.seurat$percent.ribo,probs=c(0.99)),60))

cutoffs[["percent.mito"]] <- c(min=max(quantile(data.seurat$percent.mito,probs=c(0.01)),0),max=min(quantile(data.seurat$percent.mito,probs=c(0.99)),7.5))

cutoffs[["nFeature_RNA"]] <- c(min=max(quantile(data.seurat$nFeature_RNA,probs=c(0.01)),500),max=min(quantile(data.seurat$nFeature_RNA,probs=c(0.99)),7000))
cutoffs[["nCount_RNA"]] <- c(min=max(quantile(data.seurat$nCount_RNA,probs=c(0.01)),500),max=min(quantile(data.seurat$nCount_RNA,probs=c(0.99)),50000))

print(cutoffs)

dim(data.seurat)

data.seurat <- subset(data.seurat, subset = nFeature_RNA > cutoffs[["nFeature_RNA"]]["min"] & nFeature_RNA < cutoffs[["nFeature_RNA"]]["max"] & 
                       nCount_RNA > cutoffs[["nCount_RNA"]]["min"] & nCount_RNA < cutoffs[["nCount_RNA"]]["max"] &
                       percent.ribo > cutoffs[["percent.ribo"]]["min"] &  percent.ribo < cutoffs[["percent.ribo"]]["max"] &
                    percent.mito > cutoffs[["percent.mito"]]["min"] & percent.mito < cutoffs[["percent.mito"]]["max"] )
dim(data.seurat)

Idents(data.seurat) <- "Sample"
VlnPlot(data.seurat, features = c("nFeature_RNA", "nCount_RNA","percent.ribo","percent.mito"), ncol = 2, pt.size=0)
```
Subset on singlet annotations. Cell labeling (and consequently demultiplexing) was suboptimal, so we lose a considerable amount of cells (negative or doublets)
```{r}
#exclude NA
cells <- colnames(data.seurat)[!is.na(data.seurat$Sample)]

if (length(cells)!=ncol(data.seurat)) {
   data.seurat <- subset(data.seurat, cells = cells)
}
#CMO309 corresponds to HER2_CART,  CMO309 corresponds to HER2_CART_IL10

data.seurat$condition <- as.character(data.seurat$Sample)
data.seurat$condition[data.seurat$Sample == "CMO309"] <- "WT"
data.seurat$condition[data.seurat$Sample == "CMO310"] <- "IL10"

#Remove Negatives and Doublets
data.seurat <- subset(data.seurat, condition %in% c("WT","IL10"))
Idents(data.seurat) <- "condition"
```

Load in list of genes to exclude from dim reduction (ribosomal, mitochondrial, TCR, cell cycle genes) that may distort the space
```{r}
library(scGate)
removeGenes <- unique(unique(unlist(scGate::genes.blacklist.default$Mm)))
length(removeGenes)
```


```{r}
data.seurat <- NormalizeData(data.seurat)
VlnPlot(data.seurat, features = c("Ptprc", "Cd2","Cd8a","Cd4"), ncol = 2, pt.size=0.01)
VlnPlot(data.seurat, features = c("Prf1", "Gzmb","Ifng","Pdcd1"), ncol = 2, pt.size=0.01)

data.seurat <- FindVariableFeatures(data.seurat, selection.method = "vst", nfeatures = 700, verbose = FALSE)
length(data.seurat@assays$RNA@var.features)
data.seurat@assays$RNA@var.features <- data.seurat@assays$RNA@var.features[!data.seurat@assays$RNA@var.features %in% removeGenes]
```


```{r}
set.seed(12345)
ndim=15

data.seurat <- ScaleData(data.seurat, do.scale = TRUE, do.center = TRUE) 
data.seurat <- RunPCA(object = data.seurat, features = data.seurat@assays$RNA@var.features, ndims.print = 1:5, nfeatures.print = 10, npcs = ndim)
data.seurat <- RunUMAP(data.seurat, reduction = "pca", dims = 1:ndim, seed.use=123)
```

```{r}
DimPlot(data.seurat, reduction = "umap", group.by = "condition") + ggtitle("UMAP by study")
```

Note cluster of contaminants. Probably myeloid cells (Spi1, Tyrobp, Lyz2) 

```{r fig.height=3}
FeaturePlot(data.seurat, features = c("Spi1","Tyrobp","Fcgr3","Msr1"))
FeaturePlot(data.seurat, features = c("Ptprc","Cd2","Cd3e","Cd3g"))
FeaturePlot(data.seurat, features = c("Cd4","Cd8a","Cd8b1","Foxp3"))
FeaturePlot(data.seurat, features = c("Lyz2","C1qb","Tyrobp","Trdc"))
```
Filter on T cells (remove contaminants)
```{r}
scGateModels <- get_scGateDB()

model.Tcell <- scGateModels$mouse$generic$Tcell
model.Tcell <- rbind(model.Tcell, c("level2","negative","Macrophage","Lyz2;C1qb;C1qc;Tyrobp"))

data.seurat <- scGate(data.seurat, model.Tcell, 
                      additional.signatures = list("G1S"=scGate::genes.blacklist.default$Mm$cellCycle.G1S,
                                                   "G2M"=scGate::genes.blacklist.default$Mm$cellCycle.G2M))
plot_levels(data.seurat)
data.seurat <- subset(data.seurat, subset=is.pure=="Pure")
```

Recalculate embeddings
```{r}
set.seed(12345)

ndim=10
nvar=500

palette.samples <- c("#ff7c79","#6596ff")

data.seurat <- FindVariableFeatures(data.seurat, selection.method = "vst", nfeatures = nvar*2, verbose = FALSE)
length(data.seurat@assays$RNA@var.features)
data.seurat@assays$RNA@var.features <- data.seurat@assays$RNA@var.features[!data.seurat@assays$RNA@var.features %in% removeGenes]
data.seurat@assays$RNA@var.features <- head(data.seurat@assays$RNA@var.features, nvar)

data.seurat <- ScaleData(data.seurat, do.scale = TRUE, do.center = TRUE) 
data.seurat <- RunPCA(object = data.seurat, features = data.seurat@assays$RNA@var.features, ndims.print = 1:5, nfeatures.print = 10, npcs = ndim)
data.seurat <- RunUMAP(data.seurat, reduction = "pca", dims = 1:ndim, seed.use=123)

DimPlot(data.seurat, reduction = "umap", group.by = "condition", cols=palette.samples) +  theme(aspect.ratio=1.1) + ggtitle("UMAP by condition")

ggsave("plots/umap_bysample.pdf",height=4, width=6)

#data.seurat <- subset(data.seurat, subset=condition %in% c("WT","IL10"))
```
Check expression of some genes
```{r fig.height=2.5, fig.width=3}
library(patchwork)
palette2 <- c("whitesmoke","firebrick3")
FeaturePlot(data.seurat, features = c("Spi1","Tyrobp","Fcgr3","Msr1"))
FeaturePlot(data.seurat, features = c("Ptprc","Cd2","Cd3e","Cd3g"))
FeaturePlot(data.seurat, features = c("Cd4","Cd8a","Cd8b1","Il10"))
FeaturePlot(data.seurat, features = c("nCount_RNA","nFeature_RNA","percent.ribo","percent.mito"))
```

```{r fig.height=2, fig.width=8}
p <- FeaturePlot(data.seurat, features = c("Tcf7","Il7r","Gzmb","Prf1","Pdcd1","Ifng","Ccr7","Fos","Gzmc","Gzmf","Havcr2","Mki67"), 
                 max.cutoff = 'q99', pt.size = 0.3, combine=F)

p <- FeaturePlot(data.seurat, features = c("Tcf7","Il7r","Xcl1","Slamf6","Cd200","Lef1","Ccr7","Crtam","Pdcd1","Gzmf","Havcr2","Mki67"), 
                 max.cutoff = 'q99', pt.size = 0.3, combine=F)


for (i in seq_along(p)) {
     p[[i]] <- p[[i]] + theme(aspect.ratio=1.1, axis.title=element_blank(),
        axis.text=element_blank(),
        axis.ticks=element_blank()) +
       scale_color_viridis(discrete = FALSE, option="C")
}
wrap_plots(p, ncol=6)
ggsave("plots/umap_key_genes.pdf",height=5, width=15)

```
Get signatures from mSigDB
```{r}
library(escape)

gene.sets.h <- getGeneSets(species="Mus musculus", library="H")
#names(gene.sets)

gsea.oxphos <- list("HALLMARK_OXIDATIVE_PHOSPHORYLATION" = gene.sets.h[["HALLMARK_OXIDATIVE_PHOSPHORYLATION"]]@geneIds)

gene.sets.c2 <- getGeneSets(species="Mus musculus", library="C2", 
                            gene.sets=c("KEGG_OXIDATIVE_PHOSPHORYLATION","MOOTHA_MITOCHONDRIA","BIOCARTA_ETC_PATHWAY",
                                        "KEGG_PYRUVATE_METABOLISM","REACTOME_MITOCHONDRIAL_BIOGENESIS",
                                        "KEGG_T_CELL_RECEPTOR_SIGNALING_PATHWAY","BIOCARTA_CTL_PATHWAY",
                                        "WP_TCELL_ANTIGEN_RECEPTOR_TCR_SIGNALING_PATHWAY","PID_TCR_PATHWAY"))

gsea.c2 <- lapply(gene.sets.c2, function(x){x@geneIds})
names(gsea.c2) <- names(gene.sets.c2)

gene.sets.c5 <- getGeneSets(species="Mus musculus", library="C5", 
                            gene.sets=c("GOCC_RESPIRASOME"))
gsea.c5 <- lapply(gene.sets.c5, function(x){x@geneIds})
names(gsea.c5) <- names(gene.sets.c5)


```

Evaluate multi-gene signatures
```{r fig.height=3, fig.width=6}
signatures <- list("Cytotoxic"=c("Prf1","Gzma","Gzmb","Gzmc","Gzmf"), 
                   "Stemness"=c("Tcf7","Ccr7","Sell","Lef1","Il7r"),
                   "Exhaustion"=c("Pdcd1","Lag3","Havcr2","Entpd1","Ctla4"),
                   "TCR_signaling"=c("Cd69","Cd28","Nr4a1","Nr4a2","Nr4a3","Nfkb1","Nfkbia","Nfkbib"),
                   "Cycling"=c(scGate::genes.blacklist.default$Mm$cellCycle.G1S, scGate::genes.blacklist.default$Mm$cellCycle.G2M))


#Using signatures from Carmona et al. Oncoimmunology et al.
#signatures <- list("Cytotoxic"=c("Prf1","Gzmb","Fasl"), 
#                   "Stemness"=c("Tcf7","Sell","Lef1","Il7r"),
#                   "Exhaustion"=c("Pdcd1","Lag3","Havcr2","Tigit","Ctla4"),
#                   "TCR_signaling"=c("Cd69","Cd28","Nr4a1","Nr4a2","Nr4a3","Nfkb1","Nfkbia","Nfkbib"),
#                   "Cycling"=c(scGate::genes.blacklist.default$Mm$cellCycle.G1S, scGate::genes.blacklist.default$Mm$cellCycle.G2M))


signatures <- c(signatures, gsea.oxphos, gsea.c2, gsea.c5)

data.seurat <- AddModuleScore_UCell(data.seurat, features = signatures)
signames <- paste0(names(signatures), "_UCell") 

p <- FeaturePlot(data.seurat, features = signames, 
                 max.cutoff = 'q99', pt.size = 0.5, combine=F)
       
for (i in seq_along(p)) {
     p[[i]] <- p[[i]] + theme(aspect.ratio=1.1,
        axis.title=element_blank(),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        plot.title = element_text(size=6)) +
       scale_color_viridis(discrete = FALSE, option="C")
}
wrap_plots(p, ncol=4)
ggsave("plots/umap_signatures.pdf",height=10, width=15)

#Selection
names(p) <- names(signatures)
which.sign <- c("Cytotoxic","Exhaustion","Cycling","TCR_signaling","HALLMARK_OXIDATIVE_PHOSPHORYLATION","KEGG_T_CELL_RECEPTOR_SIGNALING_PATHWAY")

pp <- p[which.sign]
wrap_plots(pp, ncol=6)
ggsave("plots/umap_signatures.select.pdf",height=4, width=16)

```

Split plots
```{r fig.width=9, fig.height=2.5}
d1 <- subset(data.seurat, subset=condition == "WT")
p1 <- FeaturePlot(d1, features = paste0(which.sign,"_UCell"), cols=palette2, max.cutoff = 'q99',
                   pt.size = 0.5, combine = F)

d2 <- subset(data.seurat, subset=condition == "IL10")
p2 <- FeaturePlot(d2, features = paste0(which.sign,"_UCell"), cols=palette2, max.cutoff = 'q99',
                   pt.size = 0.5, combine = F)

p <- c(p1, p2)

for (i in seq_along(p)) {
     p[[i]] <- p[[i]] + theme(aspect.ratio = 1.1,
        axis.title=element_blank(),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        plot.title = element_text(size=6)) +
       scale_color_viridis(discrete = FALSE, option="C")
}
wrap_plots(p, ncol=length(p)/2)

ggsave("plots/umap_signatures.splitbysample.pdf",height=6, width=16)

```

```{r}
hist(data.seurat$Cycling_UCell, breaks = 20)
data.seurat$is.cycling <- data.seurat$Cycling_UCell > 0.1
```

Unsupervised clustering
```{r}
kp <- 30
resol <- 0.35
#resol <- 0.6

#kp <- 20
#resol <- 0.3


set.seed(123)

data.seurat <- FindNeighbors(data.seurat, reduction = "pca", dims = 1:ndim, k.param = kp, nn.method = "rann")
#data.seurat <- FindNeighbors(data.seurat, reduction = "pca", dims = 1:ndim)

data.seurat  <- FindClusters(data.seurat, resolution = resol)
data.seurat$cluster <- data.seurat@active.ident

library(RColorBrewer)

display.brewer.pal(n = 8, name = 'Set2')
brewer.pal(n=8, name="Set2")

cpalette <- c("#FFD92F","#FC8D62","#B3B3B3","#8DA0CB")
#cpalette <- c("#ff7c79","#6596ff","#d574fb", "#cacaca")

g <- DimPlot(data.seurat, reduction = "umap", group.by = "cluster", pt.size = 0.5, cols=cpalette, label = T) + 
     NoLegend() + ggtitle("Unsupervised clusters") + theme(aspect.ratio=1.1)

g

ggsave("plots/unsup_clustering_res0.3.pdf", plot=g, height=4, width=6)
```

Frequency of cells by cluster and sample
```{r}
tab <- table(data.seurat$condition, data.seurat$cluster)
tab.norm <- apply(tab, 1, function(x){x/(sum(x))})

to.plot <- reshape2::melt(tab.norm, varname=c("Cluster","Condition"), value.name = "Frequency")
to.plot$Cluster <- as.character(to.plot$Cluster)

ggplot(to.plot, aes(fill=Cluster, y=Frequency, x=Condition)) + 
    geom_bar(position="stack", stat="identity") +
  theme(axis.text.x = element_text(angle = 45)) +
  scale_fill_manual(values=cpalette) +
    ggtitle("Cluster distribution by treatment") + theme_bw() + coord_flip()

ggsave("plots/unsup_cluster_frequency.flip.pdf", height = 2, width=5)

ggplot(to.plot, aes(fill=Cluster, y=Frequency, x=Condition)) + 
    geom_bar(position="stack", stat="identity") +
  theme(axis.text.x = element_text(angle = 45)) +
  scale_fill_manual(values=cpalette) + theme_bw()

ggsave("plots/unsup_cluster_frequency.pdf", height = 5, width=3)


```


Find DEG per cluster
```{r}
library(dplyr)
Idents(data.seurat) <- "cluster"

set.seed(1234)

cluster.markers <- FindAllMarkers(data.seurat, only.pos = T, min.pct = 0.1, min.diff.pct=0.1, 
                                  logfc.threshold = 0.25, max.cells.per.ident = 500, test.use="wilcox",base=exp(1))

all <- cluster.markers %>% group_by(cluster) %>% top_n(n = 50, wt = abs(avg_logFC))

for (i in levels(data.seurat@active.ident)) {
    print(subset(all, cluster==i))
}  

```


Find differentially expressed genes between clusters 0 and 1
```{r fig.height=5}
library(EnhancedVolcano)
Idents(data.seurat) <- "cluster"
markers <- FindMarkers(data.seurat, min.pct = 0.1, min.diff.pct=0.05, ident.1 = "1", ident.2 = "0",
                                  logfc.threshold = 0.05, max.cells.per.ident = 500, test.use="wilcox",base=exp(1))

EnhancedVolcano(markers, lab = rownames(markers), x = 'avg_logFC', y = 'p_val', FCcutoff=0.5, pCutoff = 10^(-5), title="Diff. expressed genes", subtitle = "Cluster 0 vs. cluster 1", drawConnectors = T)

ggsave("plots/volcano_cluster0_vs_1.pdf", height=8, width=10)
```


```{r}
feats1 <- c("Ccl5","Txnip","Pdcd4","Pmepa1","Dgat1","Slfn2") 
feats2 <- c("Gzmb","Gzmc","Gzmf","Prf1","Pdcd1","Havcr2","Cd69")
feats3 <- c("Top2a","Birc5","Mki67","Mcm3","Mcm7")
feats <- c(feats1, feats2, feats3)

DefaultAssay(data.seurat) <- "RNA"
data.seurat <- ScaleData(data.seurat, features = rownames(data.seurat@assays$RNA@data), scale.max = 2)
DoHeatmap(data.seurat, features=feats, group.by = 'cluster', group.colors = cpalette, raster=F) + 
  scale_fill_gradientn(colors = c("blue", "white", "red"))

ggsave("plots/heatmap_0123_genepanel.pdf", height = 4, width=8)
```
Define a function to plot heatmap of average expression per cluster
```{r}
make.heatmap <- function(data, assay="RNA", genes, scale="row",
                         method=c("ward.D", "average"), brewer.palette="RdBu", rev.palette=TRUE,
                         cluster.col = "cluster", flip=FALSE, min=-2, max=2,
                         palette = NULL) {
  
  library(pheatmap)
  set.seed(123)
  #Calculate mean expression by cluster
  m <- c()
  for( g in unique(genes)){
    m[[g]] <- tapply(data@assays[[assay]][g,],data[[cluster.col]], mean)
  }
  m <- as.data.frame(m)
  
  clusters = rownames(m)

  breaksList = seq(min, max, by = 0.1)
  
  require(RColorBrewer)
  p <- brewer.pal(n = 7, name = brewer.palette)
  if (rev.palette) {p <- rev(p)}
  color = colorRampPalette(p)(length(breaksList))
  
  if (is.null(palette)) {
    palette = brewer.pal(n=length(clusters), name="Paired")
    names(palette) <- clusters
  }
  
  annotation_col = data.frame(
                    cluster = clusters
                )
  rownames(annotation_col) = rownames(m)


  if (flip) { 
    h <- pheatmap::pheatmap(m, cluster_rows = F,
                            cluster_cols = F,scale = scale,
                            breaks = breaksList, color=color, 
                            annotation_row = annotation_col, 
                            show_rownames = F,
                            border_color = NA,
                            annotation_colors = list(cluster=palette), 
                            fontsize_row=6,fontsize = 7, 
                            clustering_method=method)
  } else {
    h <- pheatmap::pheatmap(t(m),cluster_rows = F,
                            cluster_cols = F,scale = scale,
                            breaks = breaksList, color=color, 
                            annotation_col = annotation_col, 
                            show_colnames = F,
                            border_color = NA,
                            annotation_colors = list(cluster=palette), 
                            fontsize_row=6,fontsize = 7, 
                            clustering_method=method)
  }
  return(h)
}
```

```{r fig.width=1.2, fig.height=1.6}
feats <- c(feats1, feats2, feats3)

names(cpalette) <- levels(data.seurat$cluster)

h <- make.heatmap(data.seurat, genes=feats, palette=cpalette, scale="row", 
                  min=-1.5, max=1.5, cluster.col = "cluster")
print(h)

ggsave("plots/heatmap_bulk_0123.vertical.pdf", plot=h, height = 3, width=2.2)
```

These plots are rescaled to min and max, therefore the expression is only relative. They can be misleading for genes that do not vary very much between clusters. How do the plots look like in absolute gene expression scale?

```{r fig.width=1.2, fig.height=1.6}
feats <- c(feats1, feats2, feats3)

names(cpalette) <- levels(data.seurat$cluster)

h <- make.heatmap(data.seurat, genes=feats, palette=cpalette, scale="none", 
                  min=0, max=4, cluster.col = "cluster", brewer.palette = "Reds", rev.palette = F)
print(h)

ggsave("plots/heatmap_bulk_0123.absolute.pdf", plot=h, height = 3, width=2.2)
```




Export object
```{r}
saveRDS(data.seurat, file="HER2_CART_seurat.rds")
```








Are there differences between WT and IL10 CARs, within the same cluster?
```{r fig.height=6, fig.width=9}
library(EnhancedVolcano)

Idents(data.seurat) <- "condition"

plots <- list()

#Cluster 0
sub <- subset(data.seurat, subset=cluster=="0")

markers <- FindMarkers(sub, min.pct = 0.1, min.diff.pct=0.05, ident.1 = "IL10", ident.2 = "WT",
                                  logfc.threshold = 0.05, max.cells.per.ident = 500, test.use="wilcox",base=exp(1))

plots[["c0"]] <- EnhancedVolcano(markers, lab = rownames(markers), x = 'avg_logFC', y = 'p_val', FCcutoff=0.5, pCutoff = 10^(-5), title="Diff. expressed genes", subtitle = "IL10 vs. WT in cluster 0", drawConnectors = T, ylim = c(0,80))

#Cluster 1
sub <- subset(data.seurat, subset=cluster=="1")

markers <- FindMarkers(sub, min.pct = 0.1, min.diff.pct=0.05, ident.1 = "IL10", ident.2 = "WT",
                                  logfc.threshold = 0.05, max.cells.per.ident = 500, test.use="wilcox",base=exp(1))

plots[["c1"]] <- EnhancedVolcano(markers, lab = rownames(markers), x = 'avg_logFC', y = 'p_val', FCcutoff=0.5, pCutoff = 10^(-5), title="Diff. expressed genes", subtitle = "IL10 vs. WT in cluster 1", drawConnectors = T, ylim = c(0,80))

#Cluster 2
sub <- subset(data.seurat, subset=cluster=="2")

markers <- FindMarkers(sub, min.pct = 0.1, min.diff.pct=0.05, ident.1 = "IL10", ident.2 = "WT",
                                  logfc.threshold = 0.05, max.cells.per.ident = 500, test.use="wilcox",base=exp(1))

plots[["c2"]] <- EnhancedVolcano(markers, lab = rownames(markers), x = 'avg_logFC', y = 'p_val', FCcutoff=0.5, pCutoff = 10^(-5), title="Diff. expressed genes", subtitle = "IL10 vs. WT in cluster 2", drawConnectors = T, ylim = c(0,80))

#Cluster 3
sub <- subset(data.seurat, subset=cluster=="3")

markers <- FindMarkers(sub, min.pct = 0.1, min.diff.pct=0.05, ident.1 = "IL10", ident.2 = "WT",
                                  logfc.threshold = 0.05, max.cells.per.ident = 500, test.use="wilcox",base=exp(1))

plots[["c3"]] <- EnhancedVolcano(markers, lab = rownames(markers), x = 'avg_logFC', y = 'p_val', FCcutoff=0.5, pCutoff = 10^(-5), title="Diff. expressed genes", subtitle = "IL10 vs. WT in cluster 3", drawConnectors = T, ylim = c(0,80))

wrap_plots(plots)

ggsave("plots/il10vsWT.volcano.bycluster.pdf", height=14, width=14)

```
